src/TrapException.cs

```diff
@@ -128,65 +273,37 @@ public class TrapException : WasmtimeException
         /// </summary>
         public int? ExitCode { get; private set; }
 
+        /// <summary>
+        /// Indentifies which type of trap this is.
+        /// </summary>
+        public TrapCode Type { get; private set; }
+
         /// <inheritdoc/>
         protected TrapException(SerializationInfo info, StreamingContext context) : base(info, context) { }
 
-        private TrapException(string message, IReadOnlyList<TrapFrame> frames) : base(message)
+        internal TrapException(string message, IReadOnlyList<TrapFrame>? frames, TrapCode type) : base(message)
         {
+            Type = type;
             Frames = frames;
         }
-        private static TrapCode GetTrapCode(IntPtr trap)
-        {
-            if (Native.wasmtime_trap_code(trap, out TrapCode code))
-            {
-                return code;
-            }
-            return TrapCode.Undefined;
-        }
 
-        internal static TrapException FromOwnedTrap(IntPtr trap)
+        internal static TrapException FromOwnedTrap(IntPtr trap, bool delete = true)
         {
-            unsafe
+            var accessor = new TrapAccessor(trap);
+            
+            var trappedException = new TrapException(accessor.Message, accessor.GetFrames(), accessor.TrapCode)
             {
-                Native.wasm_trap_message(trap, out var bytes);
-
-                var trapCode = GetTrapCode(trap);
-
-                bool trappedExit = Native.wasmtime_trap_exit_status(trap, out int exitStatus);
-
-                var byteSpan = new ReadOnlySpan<byte>(bytes.data, checked((int)bytes.size));
-
-                int indexOfNull = byteSpan.LastIndexOf((byte)0);
-                if (indexOfNull != -1)
-                {
-                    byteSpan = byteSpan.Slice(0, indexOfNull);
-                }
+                ExitCode = accessor.ExitStatus
+            };
 
-                var message = Encoding.UTF8.GetString(byteSpan);
-                bytes.Dispose();
+            if (delete)
```

Comment:
Nit: could we put braces around this statement?

Additionally, I wonder if we should wrap the above in a `try`..`finally` to dispose the trap accessor (when `delete` is true) just so we don't accidentally leak the trap if something goes wrong.